### PR TITLE
#11207: Make Device type opaque

### PR DESCRIPTION
**Pull request by patrickroberts**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/11207

### Problem description
This work is intended as a proposal in order to formalize the TT Metal API, Metal API types becoming opaque.  
Should this be accepted, the following tasks would be updated similarly.

- [x] #11211
- [x] #11210
- [x] #11208

### What's changed
Add formal API methods for member methods of tt_metal::Device so that it can be an incomplete/opaque type.

1. Moved Device internals to device_impl.hpp
The Device class's internal implementation, including data members and private methods, was moved to a new file device_impl.hpp. This allows the class definition to remain hidden from the public API while still being accessible internally.

2. Introduced forward declaration of Device in public header.
This ensures that the class is incomplete in the public API, with no direct access to its internal structure.

3. API Refactoring
Functions interacting with Device were refactored to use opaque pointers. 
For example:
- ARCH DeviceArch(const Device *device);
- chip_id_t DeviceId(const Device *device);
- bool DeviceIsInitialized(Device *device);

Device internals are now hidden via API functions instead of directly exposing the internal structure. 

4. Refactor Internal Calls to Use device_impl.hpp
All internal calls and logic dealing with the Device internals were updated to use the newly relocated device_impl.hpp. Any existing direct access in public headers were replaced with calls to the refactored API.

5. Replaced arguments with aggregate types.   
For example, CreateDevice now uses the DeviceOptions struct to simplify the CreateDevice process. Instead of passing individual parameters, the configuration options are encapsulated in DeviceOptions:

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by patrickroberts** [comment]
This is currently failing because pybind11 can't get the `typeid` of `Device` unless it's a complete type... will look into this later.

File: tt_metal/impl/device/device_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+    void enable_async(bool enable);
+    WorkExecutorMode get_worker_mode() { return work_executor.get_worker_mode(); }
+    void set_worker_queue_mode(const WorkerQueueMode& mode) { this->work_executor.set_worker_queue_mode(mode); }
+    WorkerQueueMode get_worker_queue_mode() { return this->work_executor.get_worker_queue_mode(); }
+    // TODO: Uplift usage of friends. Buffer and Program just need access to allocator
+    friend class Buffer;
+    friend class Program;
+    friend class SystemMemoryManager;
+
+    static constexpr MemoryAllocator allocator_scheme_ = MemoryAllocator::L1_BANKING;
```
Comment (ayerofieiev-tt) [inline tt_metal/impl/device/device_impl.hpp:216]: can members go to private? or not yet? Not that opaque :D

Comment (patrickroberts) [reply tt_metal/impl/device/device_impl.hpp:216]: It's opaque in the sense that nothing outside of the TT Metal library sees its complete type. Changing the class definition in any way is out of scope for this PR.

**Review by ayerofieiev-tt — COMMENTED** [review]
Lots of work! Thank you, Patrick! 

--------------

The change affects important public facing API. 
Plans for such changes must be communicated to users/customers via github discussions / discord / private channels ahead of time.

@davorchap , @eyonland is anyone working on the community announcement?

File: tt_metal/impl/device/device.hpp [hunk 2]
```diff
@@ -62,321 +62,75 @@ static constexpr float  INF_WHB0 = 1.7014e+38;
...
-        const uint8_t num_hw_cqs,
-        std::size_t l1_small_size,
-        std::size_t trace_region_size,
-        const std::vector<uint32_t> &l1_bank_remap = {},
-        bool minimal = false,
-        uint32_t worker_core = 0);
+class Device;
 
-    ~Device();
+ARCH DeviceArch(const Device *device);
```
Comment (pavlejosipovic) [inline tt_metal/impl/device/device.hpp:67]: Do want a namespace to group these device apis?

**Comment by razorback3** [comment]
FYI. @namhyeong-kim

**Comment by eyonland** [comment]
> Lots of work! Thank you, Patrick!
> 
> The change affects important public facing API. Plans for such changes must be communicated to users/customers via github discussions / discord / private channels ahead of time.
> 
> @davorchap , @eyonland is anyone working on the community announcement?

This is a draft PR intended for internal discussions only at this point.  Please allow us time to review this impact and plan our approach.

**Review by pgkeller — COMMENTED** [review]
Some thoughts:
1) I assume the void fns returning values is just a search/replace artifact of the prototype change
2) I assume all the top-level fns that take a device are placeholders for the "reach around" APIs
3) Before we merge this, I'd like to whack at the list of reach arounds and reduce the total.  They don't all have to be done, but it looks like quite a few could be consolidated w/ at least the "get/query" API, eg, DeviceArch, DeviceL1SizePerCore, Grid sizes, etc)
4) I'd like to split out the API implementation from the device internal implementation (eg, line 114->116 in device.cpp). I think a dir tt_metal/api or some such makes sense

**Pull request closed without merging**